README
# Duration representation for humans!

###### \[in package HUMANIZE-DURATION with nicknames HUMANIZE-DURATION/CORE\]
This is a small library usefult for time duration humanization.

## Introduction

It is different from LOCAL-TIME-DURATION:HUMAN-READABLE-DURATION, because allows
to output only significant parts of a duration object.

Also it limits a number of part. For example, if there days, hours, minutes and seconds,
then most probably it is already not important exactly how many minutes and seconds
passed since the moment.

Compare these two examples:

### local-time-duration:human-readable-duration

```
CL-USER> (local-time-duration:human-readable-duration
          (local-time-duration:duration :day 5
                                        :hour 13
                                        :minute 0
                                        :sec 14
                                        :nsec 10042))
" 5 days 13 hours 14 seconds 10042 nsecs"
```

### humanize-duration:humanize-duration

```
CL-USER> (humanize-duration:humanize-duration
          (local-time-duration:duration :day 5
                                        :hour 13
                                        :minute 0
                                        :sec 14
                                        :nsec 10042))
"5 days 13 hours"
```

Main job is done at HUMANIZE-DURATION:

- [function] HUMANIZE-DURATION DURATION &KEY STREAM (N-PARTS 2) (FORMAT-PART #'DEFAULT-FORMAT-PART)

    This is the better version of LOCAL-TIME-DURATION:HUMAN-READABLE-DURATION.
    
    By default it returns only 2 most significant duration parts.
    
    If duration is 2 hour, 43 seconds and 15 nanoseconsds, then
    function will return "2 hours 43 seconds":
    
    ```lisp
    CL-USER> (ultralisp/utils/time:humanize-duration
              (local-time-duration:duration :hour 2
                                            :sec 43
                                            :nsec 15))
    "2 hours 43 seconds"
    ```
    
    Also, you can pass a `:format-part` argument.
    It should be a function of three arguments:
    `(stream part-type part)` where `part-type` is a keyword
    from this list:
    
    ```lisp
    (list :weeks :days :hours :minutes :secs :nsecs)
    ```


HUMANIZE-DURATION accepts :FORMAT-PART argument, which is DEFAULT-FORMAT-PART function by default:
your own version. This could be useful if you want to support localization to other languages.

- [function] DEFAULT-FORMAT-PART STREAM PART-TYPE PART

    This is should return a string with propertly pluralized form.
    
    - PART-TYPE argument is a member of (list :weeks :days :hours :minutes :secs :nsecs).
    
    - PART is an integer.
    
    Here are possible results:
    
        (t :weeks 1) -> "1 week"
        (t :weeks 5) -> "5 weeks"
        (t :day 2) -> "2 days"


Here is how you can localize output for your language

## Localization

HUMANIZE-DURATION comes with predefined Russian localization.

### Russian localization

###### \[in package HUMANIZE-DURATION/RU\]
This package includes a single function, useful to display duration in Russian language:

- [function] FORMAT-PART STREAM PART-TYPE PART

    This is Russian version of part formatter for HUMANIZE-DURATION

You can use it as an template, to define other languages.

Please, don't forget to contribute your solutions!

Here is how you can use it in your code:

```
CL-USER> (humanize-duration:humanize-duration
          (local-time-duration:duration :day 5
                                        :hour 13
                                        :minute 0
                                        :sec 14
                                        :nsec 10042)
          :format-part #'humanize-duration/ru:format-part)
"5 дней 13 часов"
```

Here is how FORMAT-PART is defined in the code:

```commonlisp
(defun format-part (stream part-type part)
  "This is Russian version of part formatter for HUMANIZE-DURATION"
  (format stream "~d ~A"
          part
          (apply #'choose-form
                 part
                 (ecase part-type
                   (:weeks '("неделя" "недели" "недель"))
                   (:days '("день" "дня" "дней"))
                   (:hours '("час" "часа" "часов"))
                   (:minutes '("минута" "минуты" "минут"))
                   (:secs '("секунда" "секунды" "секунд"))
                   (:nsecs '("наносекунда" "наносекунды" "наносекунд"))))))

(defvar %end-of-format-part%)
```

Russian version uses internal helper, to choose a correct word form:

```commonlisp
(defun choose-form (n &rest forms)
  "This function is bassed on this gettext formula:

   ```
   Plural-Forms: nplurals=3; plural=n%10==1 && n%100!=11 ? 0 : n%10>=2 && n%10<=4 && (n%100<10 || n%100>=20) ? 1 : 2;
   ```
"
  (let ((n%10 (rem n 10))
        (n%100 (rem n 100)))
    (cond
      ((and (= n%10
               1)
            (not (= n%100
                    11)))
       (first forms))
      ((and (>= n%10
                2)
            (<= n%10
                4)
            (or (<= n%100
                    10)
                (>= n%100
                    20)))
       (second forms))
      (t
       (third forms)))))

```

Applied to a different numbers it produces the following:

```
CL-USER> (flet ((p (n)
                  (format t "~A ~A~%"
                          n
                          (choose-form n "яблоко" "яблока" "яблок"))))
           (loop for i upto 12
                 do (p i)))
0 яблок
1 яблоко
2 яблока
3 яблока
4 яблока
5 яблок
6 яблок
7 яблок
8 яблок
9 яблок
10 яблок
11 яблок
12 яблок
```


* * *
###### \[generated by [MGL-PAX](https://github.com/melisgl/mgl-pax)\]


ChangeLog.md
<a id="x-28HUMANIZE-DURATION-2FCHANGELOG-3A-40CHANGELOG-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

# ChangeLog

<a id="x-28HUMANIZE-DURATION-2FCHANGELOG-3A-3A-7C0-2E2-2E0-7C-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

## 0.2.0 (2021-09-11)

Move to a new documentation renderer.

<a id="x-28HUMANIZE-DURATION-2FCHANGELOG-3A-3A-7C0-2E1-2E0-7C-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

## 0.1.0 (2021-01-25)

Initial version.


* * *
###### [generated by [40ANTS-DOC](https://40ants.com/doc/)]


src/ru.lisp
(uiop:define-package #:humanize-duration/ru
  (:use #:cl)
  (:import-from #:40ants-doc
                #:defsection)
  (:export
   #:format-part))
(in-package humanize-duration/ru)


(defsection @index (:title "Russian localization")
  "This package includes a single function, useful to display duration in Russian language:"
  
  (format-part function)

  "You can use it as an template, to define other languages.

   Please, don't forget to contribute your solutions!"

  "Here is how you can use it in your code:

   ```
   CL-USER> (humanize-duration:humanize-duration
             (local-time-duration:duration :day 5
                                           :hour 13
                                           :minute 0
                                           :sec 14
                                           :nsec 10042)
             :format-part #'humanize-duration/ru:format-part)
   \"5 дней 13 часов\"
   ```

   Here is how FORMAT-PART function is defined in the code:
"
  (format-part (include (:start (format-part function)
                         :end (%end-of-format-part% variable))))
  
  "Russian version uses internal helper, to choose a correct word form:"

  (choose-form (include (:start (choose-form function)
                         :end (%end-of-choose-form% variable))))

  "Applied to a different numbers it produces the following:

   ```
   CL-USER> (flet ((p (n)
                     (format t \"~A ~A~%\"
                             n
                             (choose-form n \"яблоко\" \"яблока\" \"яблок\"))))
              (loop for i upto 12
                    do (p i)))
   0 яблок
   1 яблоко
   2 яблока
   3 яблока
   4 яблока
   5 яблок
   6 яблок
   7 яблок
   8 яблок
   9 яблок
   10 яблок
   11 яблок
   12 яблок
   ```
")


(defun choose-form (n &rest forms)
  "This function is based on this gettext formula:

   ```
   Plural-Forms: nplurals=3; plural=n%10==1 && n%100!=11 ? 0 : n%10>=2 && n%10<=4 && (n%100<10 || n%100>=20) ? 1 : 2;
   ```
"
  (let ((n%10 (rem n 10))
        (n%100 (rem n 100)))
    (cond
      ((and (= n%10
               1)
            (not (= n%100
                    11)))
       (first forms))
      ((and (>= n%10
                2)
            (<= n%10
                4)
            (or (<= n%100
                    10)
                (>= n%100
                    20)))
       (second forms))
      (t
       (third forms)))))

(defvar %end-of-choose-form%)


(defun format-part (stream part-type part)
  "This is Russian version of part formatter for HUMANIZE-DURATION"
  (format stream "~d ~A"
          part
          (apply #'choose-form
                 part
                 (ecase part-type
                   (:weeks '("неделя" "недели" "недель"))
                   (:days '("день" "дня" "дней"))
                   (:hours '("час" "часа" "часов"))
                   (:minutes '("минута" "минуты" "минут"))
                   (:secs '("секунда" "секунды" "секунд"))
                   (:nsecs '("наносекунда" "наносекунды" "наносекунд"))))))

(defvar %end-of-format-part%)


src/app.lisp
(defpackage #:example/app
  (:use #:cl)
  (:import-from #:example/utils)
  (:import-from #:mgl-pax
                #:defsection)
  (:documentation "This is docstring for the package.

                   The package contains a function which does it's job by
                   applying transformation to the first and second arguments.

                   Note, despite the docstring indentation, it is displayed
                   correctly. And PAX is smart enough to distinguish the common
                   indentation and indentation of a code block below:

                       (apply #'foo 42)
                  "))
(in-package example/app)


(defsection @app (:title "EXAMPLE/APP Package")
  "This is a test package of our example.

   It contains only one function definition:"
  
  (foo function)
  
  "What is interesting is that we can continue writing code for the section.
   Interleaving blocks of text with references to the different Lisp entities.

   At some moment we can reference objects from other packages. But to make them
   cross-referenced, we have to mention them in some section. For example, here FOO
   references EXAMPLE/UTILS:DO-THE-JOB, which is desribed in the EXAMPLE/UTILS:@UTILS.")


(defun foo (first &key (other 100500))
  "This is example function.

   Internally it calls [example/utils:do-the-job][function]
   to do the real job."
  (example/utils:do-the-job first other))


src/utils.lisp
(defpackage #:example/utils
  (:use #:cl)
  (:import-from #:mgl-pax
                #:defsection)
  (:export #:do-the-job)
  (:documentation "The utils"))
(in-package example/utils)


(defsection @utils (:title "EXAMPLE/UTILS Package")
  (do-the-job function))


(defun concat (first second)
  "This function is not exported and should not be showed in the API reference."
  (format nil "~A ~A" first second))


(defun do-the-job (first second)
  "The function does the job.

   It **concatenates** first and second arguments
   calling internal function concat.

   On this multiline we'll check how does documentation
   system processes docstrings.

   By the way, pay attention at the second paragraph where
   I've used [Markdown](https://www.markdownguide.org/basic-syntax/)
   format to make the word \"concatenates\" bold.

   Also, we can reference some parts of the documentation.
   Read more about cross referencing in the EXAMPLE-DOCS:@HANDWRITTEN chapter."
  (concat first second))


src/core.lisp
(uiop:define-package :humanize-duration
  (:use #:cl)
  (:nicknames :humanize-duration/core)
  (:import-from #:local-time-duration)
  (:import-from #:humanize-duration/ru)
  (:import-from #:40ants-doc
                #:defsection
                #:defsection-copy)
  (:import-from #:with-output-to-stream
                #:with-output-to-stream)
  (:import-from #:humanize-duration/changelog
                #:@changelog)
  (:import-from #:docs-config
                #:docs-config)
  (:export #:@index #:@readme
           #:default-format-part
           #:humanize-duration))
(in-package humanize-duration)


(defmethod docs-config ((system (eql (asdf:find-system "humanize-duration"))))
  ;; 40ANTS-DOC-THEME-40ANTS system will bring
  ;; as dependency a full 40ANTS-DOC but we don't want
  ;; unnecessary dependencies here:
  (ql:quickload :40ants-doc-theme-40ants)
  (list :theme
        (find-symbol "40ANTS-THEME"
                     (find-package "40ANTS-DOC-THEME-40ANTS"))))


(defsection @index (:title "Duration representation for humans!"
                    :ignore-words ("GIT"
                                   "LOCAL-TIME-DURATION:DURATION"
                                   "LOCAL-TIME-DURATION:HUMAN-READABLE-DURATION"))
  "This is a small library useful for time duration humanization."

  (humanize-duration system)
  (@intro section)
  (@localization section))

(defsection-copy @readme @index)


(defsection @intro (:title "Introduction")
  "It is different from LOCAL-TIME-DURATION:HUMAN-READABLE-DURATION, because allows
   to output only significant parts of a duration object.

   Also it limits a number of part. For example, if there days, hours, minutes and seconds,
   then most probably it is already not important exactly how many minutes and seconds
   passed since the moment.

   Compare these two examples:

   ### local-time-duration:human-readable-duration

   ```
   CL-USER> (local-time-duration:human-readable-duration
             (local-time-duration:duration :day 5
                                           :hour 13
                                           :minute 0
                                           :sec 14
                                           :nsec 10042))
   \" 5 days 13 hours 14 seconds 10042 nsecs\"
   ```

   ### humanize-duration:humanize-duration

   ```
   CL-USER> (humanize-duration:humanize-duration
             (local-time-duration:duration :day 5
                                           :hour 13
                                           :minute 0
                                           :sec 14
                                           :nsec 10042))
   \"5 days 13 hours\"
   ```

   Main job is done at HUMANIZE-DURATION:
"
  (humanize-duration function)

  "HUMANIZE-DURATION accepts :FORMAT-PART argument, which is DEFAULT-FORMAT-PART function by default:
   your own version. This could be useful if you want to support localization to other languages."

  (default-format-part function)

  "Here is how you can localize output for your language")


(defsection @localization (:title "Localization")
  "HUMANIZE-DURATION comes with predefined Russian localization."
  (humanize-duration/ru::@index section))


(defun default-format-part (stream part-type part)
  "This is should return a string with propertly pluralized form.

   - PART-TYPE argument is a member of (list :weeks :days :hours :minutes :secs :nsecs).
   - PART is an integer.

   Here are possible results:

       (t :weeks 1) -> \"1 week\"
       (t :weeks 5) -> \"5 weeks\"
       (t :day 2) -> \"2 days\"
"
  (ecase part-type
    (:weeks (format stream "~d week~:p" part))
    (:days (format stream "~d day~:p" part))
    (:hours (format stream "~d hour~:p" part))
    (:minutes (format stream "~d minute~:p" part))
    (:secs (format stream "~d second~:p" part))
    (:nsecs (format stream "~d nsec~:p" part))))


(defun humanize-duration (duration &key stream (n-parts 2) (format-part #'default-format-part))
  "This is the better version of LOCAL-TIME-DURATION:HUMAN-READABLE-DURATION.

   By default it returns only 2 most significant duration parts.

   If duration is 2 hour, 43 seconds and 15 nanoseconsds, then
   function will return \"2 hours 43 seconds\":

   ```lisp
   CL-USER> (ultralisp/utils/time:humanize-duration
             (local-time-duration:duration :hour 2
                                           :sec 43
                                           :nsec 15))
   \"2 hours 43 seconds\"
   ```

   Also, you can pass a `:format-part` argument.
   It should be a function of three arguments:
   `(stream part-type part)` where `part-type` is a keyword
   from this list:

   ```lisp
   (list :weeks :days :hours :minutes :secs :nsecs)
   ```"
  (check-type duration local-time-duration:duration)
  (check-type n-parts (integer 1 6))
  
  (multiple-value-bind (nsecs secs minutes hours days weeks)
      (local-time-duration::decode-duration duration :weeks t)
    (with-output-to-stream (stream stream)
      (let ((part-types (list :weeks :days :hours :minutes :secs :nsecs))
            (parts (list weeks days hours minutes secs nsecs)))
        (if (every #'zerop parts)
            (format stream "0 length")
            (loop with n-printed = 0
                  for part in parts
                  for part-type in part-types
                  unless (zerop part)
                  do (unless (zerop n-printed)
                       ;; Add a space between parts
                       (format stream " "))
                     (funcall format-part
                              stream
                              part-type
                              part)
                     (incf n-printed)
                  when (>= n-printed n-parts)
                  do (return)))))))
